README.md
# confluent_schema_registry

Elixir client for the [Confluent® Schema Registry](https://www.confluent.io/confluent-schema-registry).

It implements the full [REST API](https://docs.confluent.io/current/schema-registry/develop/api.html) using
the [Tesla](https://github.com/teamon/tesla) HTTP client library. This gives it support for
HTTP authentication, SSL/TLS and other configuration flexibility, e.g.
selecting the underlying HTTP library (e.g. Hackney).

Thanks to [Schemex](https://hex.pm/packages/schemex) and [Avrora](https://github.com/Strech/avrora)
for inspiration.

## Installation

Add `confluent_schema_registry` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:confluent_schema_registry, "~> 0.1.1"}]
end
```

Then run `mix deps.get` to fetch the new dependency.

Documentation is on [HexDocs](https://hexdocs.pm/confluent_schema_registry).
To generate a local copy, run `mix docs`.

### Optional dependencies

By default, Tesla uses the [httpc](http://erlang.org/doc/man/httpc.html) HTTP client
which comes with OTP. That library is pretty bare bones, and has issues with
e.g. validating SSL certificates, so I recommend using [hackney](https://hex.pm/packages/hackney).
Configure it as documented in [Tesla.Adapter.Hackney](https://hexdocs.pm/tesla/Tesla.Adapter.Hackney.html).

```elixir
config :tesla, :adapter, Tesla.Adapter.Hackney
```

[AvroSchema](https://github.com/cogini/avro_schema) provides convenience functions
to work  with Avro schemas, tag ids, encode/decode data, and cache results for performance
and reliablilty.

## Usage

First create a client:

```elixir
client = ConfluentSchemaRegistry.client(base_url: "https://schemaregistry.example.com:8081/")
```

In the Kafka [wire format](https://docs.confluent.io/current/schema-registry/serializer-formatter.html#wire-format),
the data is prefixed by magic byte 0 to indicate that it is using the schema
registry, then four bytes for the schema id, then the data.

### Producer

On startup, a producer looks up the version of the schema matching the
subject associated with the topic it is writing on. It encodes the data to
binary format using the schema, then appends the schema id to the binary and
sends it to Kafka.

There are a couple of different options for how to get the schema, depending
on the policy and permissions for updating schemas.

`is_registered/3` checks if a schema has already been registered under the
specified subject. If so, it returns the registration.

```elixir
 schema = "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}"
case ConfluentSchemaRegistry.is_registered(client, "test", schema) do
  {:ok, reg} ->
    # Found
    schema = reg["schema"]
    schema_id = reg["id"]
  {:error, 404, %{"error_code" => 40401}} ->
    # Subject not found
  {:error, 404, %{"error_code" => 40403}} ->
    # Schema not found
  {:error, code, reason} ->
    # Other error
end
```

If the schema hasn't been registered, then the producer can attempt to
register it using `register_schema/3`. If it is successful, it returns the
schema id. It can be called multiple times, and will return the same schema id.

```elixir
case ConfluentSchemaRegistry.register_schema(client, "test", schema) do
  {:ok, schema_id} ->
    # Already registered
  {:error, 409, reason} ->
    # Conflict -- Incompatible Avro schema
  {:error, 422, reason} ->
    # Unprocessable Entity, Invalid Avro schema
  {:error, code, reason} ->
    # Other error
end
```

Another option is to use `get_schema/3` to read the latest registered schema
for the subject. You might do this when schema registrations are manually managed.

```elixir
case ConfluentSchemaRegistry.get_schema(client, "test", "latest") do
  {:ok, reg} ->
    # Already registered
    schema = reg["schema"]
    schema_id = reg["id"]
  {:error, 404, %{"error_code" => 40401}} ->
    # Subject not found
  {:error, 404, %{"error_code" => 40402}} ->
    # Version not found
  {:error, 422, reason} ->
    # Unprocessable Entity, Invalid Avro version
  {:error, code, reason} ->
    # Other error
end
```

### Consumer

A consumer receives data, gets the schema id from the prefix, and looks it up
in the registry using `get_schema/1`, getting the schema which was used to write it.
It then decodes the binary using the schema.

```elixir
{:ok, schema} = ConfluentSchemaRegistry.get_schema(client, 21)
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed
- Use `Config` instead of `Mix.Config`

## [0.1.1] - 2022-22-22

### Changed
- Updated dependencies

## [0.1.0] - 2019-12-31
### Added
- Initial release


config/test.exs
import Config

# config :logger, level: :debug
# config :logger,
#   level: :debug,
#   backends: [:console],
#   compile_time_purge_level: :debug


test/cache_test.exs
defmodule CacheTest do

  use ExUnit.Case

  @moduletag :live

  # Test cache

  setup do
    # {:ok, _pid} = ConfluentSchemaRegistry.Cache.start_link([dets_dir: "/tmp"])
    {:ok, _pid} = ConfluentSchemaRegistry.Cache.start_link([])

    # client = ConfluentSchemaRegistry.client(middleware: [Tesla.Middleware.Logger])
    client = ConfluentSchemaRegistry.client()

    # Tesla.Mock doesn't work because it uses a different process for the GenServer
    # mock fn
    #   %{method: :get, url: "http://localhost:8081/schemas/ids/1"} ->
    #     schema = "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}"
    #     json(%{"schema" => schema})
    #   %{method: :get, url: "http://localhost:8081/subjects/test/versions/latest"} ->
    #     schema = "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}"
    #     json(%{"id" => 1, "schema" => schema, "subject" => "test", "version" => 1})
    #   %{method: :get, url: "http://localhost:8081/subjects/test/versions/1"} ->
    #     schema = "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}"
    #     json(%{"id" => 1, "schema" => schema, "subject" => "test", "version" => 1})
    # end

    {:ok, client: client}
  end

  test "cache", %{client: client} do
    schema = "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}"

    assert {:ok, schema} == ConfluentSchemaRegistry.Cache.get_schema(client, 21)

    cache = ConfluentSchemaRegistry.Cache.dump()
    refute Enum.empty?(cache)


    key = {ConfluentSchemaRegistry, :get_schema, [client, 21]}
    assert {:infinity, :infinity, schema} == ConfluentSchemaRegistry.Cache.cache_lookup(key)

    {:ok, versions} = ConfluentSchemaRegistry.get_versions(client, "test")
    {:ok, result} = ConfluentSchemaRegistry.Cache.get_schema(client, "test", List.last(versions))
    assert result["schema"] == schema
    assert result["subject"] == "test"

    key = {ConfluentSchemaRegistry, :get_schema, [client, "test", List.last(versions)]}
    assert {:infinity, :infinity, result} == ConfluentSchemaRegistry.Cache.cache_lookup(key)


    {:ok, result} = ConfluentSchemaRegistry.Cache.get_schema(client, "test")
    assert result["schema"] == schema
    assert result["subject"] == "test"

    key = {ConfluentSchemaRegistry, :get_schema, [client, "test", "latest"]}
    {_expires, ttl, _result} = ConfluentSchemaRegistry.Cache.cache_lookup(key)
    assert ttl == 3600

    {:ok, result} = ConfluentSchemaRegistry.Cache.is_registered(client, "test", schema)
    assert result["schema"] == schema
    assert result["subject"] == "test"

  end
end



test/client_test.exs
defmodule ClientTest do

  use ExUnit.Case

  import Tesla.Mock

  # These are the test cases from the documentation
  # https://docs.confluent.io/current/schema-registry/develop/api.html

  setup do
    client = ConfluentSchemaRegistry.client(adapter: Tesla.Mock)

    mock fn
      %{method: :get, url: "http://localhost:8081/schemas/ids/1"} ->
        json(%{"schema" => "{\"type\": \"string\"}"})
      %{method: :get, url: "http://localhost:8081/subjects"} ->
        json(["subject1", "subject2"])
      %{method: :get, url: "http://localhost:8081/subjects/test/versions"} ->
        json([1, 2, 3, 4])
      %{method: :delete, url: "http://localhost:8081/subjects/test"} ->
        json([1, 2, 3, 4])
      %{method: :get, url: "http://localhost:8081/subjects/test/versions/1"} ->
        json(%{"name" => "test", "version" => 1, "schema" => "{\"type\": \"string\"}"})
      %{method: :get, url: "http://localhost:8081/subjects/test/versions/latest"} ->
        json(%{"name" => "test", "version" => 1, "schema" => "{\"type\": \"string\"}"})
      %{method: :get, url: "http://localhost:8081/subjects/test/versions/1/schema"} ->
        %Tesla.Env{status: 200, body: "{\"type\": \"string\"}"}
      %{method: :post, url: "http://localhost:8081/subjects/test/versions"} ->
        json(%{"id" => 1})
      %{method: :post, url: "http://localhost:8081/subjects/test"} ->
        response = %{
           "subject" => "test",
           "id" => 1,
           "version" => 3,
           "schema" => """
           {
             \"type\": \"record\",
             \"name\": \"test\",
             \"fields\":
               [
                 {
                   \"type\": \"string\",
                   \"name\": \"field1\"
                 },
                 {
                   \"type\": \"int\",
                   \"name\": \"field2\"
                 }
               ]
           }
           """
        }
        json(response)
      %{method: :delete, url: "http://localhost:8081/subjects/test/versions/1"} ->
        %Tesla.Env{status: 200, body: 1}
      %{method: :delete, url: "http://localhost:8081/subjects/test/versions/latest"} ->
        %Tesla.Env{status: 200, body: 1}
      %{method: :post, url: "http://localhost:8081/compatibility/subjects/test/versions/latest"} ->
        json(%{"is_compatible" => true})
      %{method: :post, url: "http://localhost:8081/compatibility/subjects/test/versions/1"} ->
        json(%{"is_compatible" => true})
      %{method: :put, url: "http://localhost:8081/config"} ->
        json(%{"compatibility" => "FULL"})
      %{method: :get, url: "http://localhost:8081/config"} ->
        json(%{"compatibilityLevel" => "FULL"})
      %{method: :put, url: "http://localhost:8081/config/test"} ->
        json(%{"compatibility" => "FULL"})
      %{method: :get, url: "http://localhost:8081/config/test"} ->
        json(%{"compatibilityLevel" => "FULL"})
    end

    {:ok, client: client}
  end

  test "get_schema id", %{client: client} do
    assert {:ok, "{\"type\": \"string\"}"} == ConfluentSchemaRegistry.get_schema(client, 1)
  end

  test "get_subjects", %{client: client} do
    assert {:ok, ["subject1", "subject2"]} == ConfluentSchemaRegistry.get_subjects(client)
  end

  test "get_versions", %{client: client} do
    assert {:ok, [1, 2, 3, 4]} == ConfluentSchemaRegistry.get_versions(client, "test")
  end

  test "delete_subject", %{client: client} do
    {:ok, [1, 2, 3, 4]} = ConfluentSchemaRegistry.delete_subject(client, "test")
  end

  test "get_schema subject", %{client: client} do
    schema = %{"name" => "test", "version" => 1, "schema" => "{\"type\": \"string\"}"}
    {:ok, schema2} = ConfluentSchemaRegistry.get_schema(client, "test", 1)
    assert schema == schema2

    {:ok, schema3} = ConfluentSchemaRegistry.get_schema(client, "test")
    assert schema == schema3
  end

  test "register_schema", %{client: client} do
    {:ok, 1} = ConfluentSchemaRegistry.register_schema(client, "test", "schema")
  end

  test "is_registered", %{client: client} do
    response = %{"subject" => "test", "id" => 1, "version" => 3, "schema" => """
      {
        \"type\": \"record\",
        \"name\": \"test\",
        \"fields\":
          [
            {
              \"type\": \"string\",
              \"name\": \"field1\"
            },
            {
              \"type\": \"int\",
              \"name\": \"field2\"
            }
          ]
      }
      """
    }
    assert {:ok, response} == ConfluentSchemaRegistry.is_registered(client, "test", "schema")
  end

  test "delete_version", %{client: client} do
    assert {:ok, 1} == ConfluentSchemaRegistry.delete_version(client, "test") # latest
    assert {:ok, 1} == ConfluentSchemaRegistry.delete_version(client, "test", 1)
  end

  test "is_compatible", %{client: client} do
    assert {:ok, true} == ConfluentSchemaRegistry.is_compatible(client, "test", "schema") # latest
    assert {:ok, true} == ConfluentSchemaRegistry.is_compatible(client, "test", "schema", 1)
  end

  test "update_compatibility", %{client: client} do
    assert {:ok, "FULL"} == ConfluentSchemaRegistry.update_compatibility(client, "FULL")
  end

  test "get_compatibility", %{client: client} do
    assert {:ok, "FULL"} == ConfluentSchemaRegistry.get_compatibility(client)
  end

  test "update_compatibility subject", %{client: client} do
    assert {:ok, "FULL"} == ConfluentSchemaRegistry.update_compatibility(client, "test", "FULL")
  end

  test "get_compatibility subject", %{client: client} do
    assert {:ok, "FULL"} == ConfluentSchemaRegistry.get_compatibility(client, "test")
  end

end


mix.exs
defmodule ConfluentSchemaRegistry.MixProject do
  use Mix.Project

  @github "https://github.com/cogini/confluent_schema_registry"

  def project do
    [
      app: :confluent_schema_registry,
      version: "0.1.1",
      elixir: "~> 1.8",
      elixirc_paths: elixirc_paths(Mix.env()),
      start_permanent: Mix.env() == :prod,
      description: description(),
      package: package(),
      source_url: @github,
      homepage_url: @github,
      docs: docs(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],
      dialyzer: [
        # plt_add_apps: [:erlavro, :tesla],
        # plt_add_deps: true,
        # flags: ["-Werror_handling", "-Wrace_conditions"],
        flags: ["-Wunmatched_returns", :error_handling, :race_conditions, :underspecs]
        # ignore_warnings: "dialyzer.ignore-warnings"
      ],
      deps: deps()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger] ++ extra_applications(Mix.env())
    ]
  end

  defp extra_applications(:test), do: [:hackney]
  defp extra_applications(_), do: []

  # Specifies which paths to compile per environment
  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:credo, "~> 1.1", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 0.5", only: [:dev, :test], runtime: false},
      {:ex_doc, "~> 0.28", only: :dev, runtime: false},
      {:excoveralls, "~> 0.12.0", only: [:dev, :test], runtime: false},
      {:hackney, "~> 1.14", only: [:dev, :test]},
      {:jason, "~> 1.0"},
      {:telemetry, "~> 0.4 or ~> 1.0"},
      {:tesla, "~> 1.3"}
    ]
  end

  defp description do
    "Client for Confluent Schema Registry"
  end

  defp package do
    [
      maintainers: ["Jake Morrison"],
      licenses: ["Apache-2.0"],
      links: %{"GitHub" => @github}
    ]
  end

  defp docs do
    [
      source_url: @github,
      extras: ["README.md", "CHANGELOG.md"],
      # api_reference: false,
      source_url_pattern: "#{@github}/blob/master/%{path}#L%{line}"
    ]
  end
end


lib/confluent_schema_registry.ex
defmodule ConfluentSchemaRegistry do
  @moduledoc """
  Elixir client for [Confluent Schema Registry](https://docs.confluent.io/current/schema-registry/)
  [API](https://docs.confluent.io/current/schema-registry/develop/api.html).
  """

  # Argument types
  @type id :: pos_integer
  @type schema :: binary
  @type subject :: binary
  @type code :: non_neg_integer
  @type reason :: any
  @type version :: pos_integer | binary
  @type level :: binary

  @doc ~S"""
  Create client to talk to Schema Registry.

  Options are:

  * base_url: URL of schema registry (optional, default "http://localhost:8081")
  * username: username for BasicAuth (optional)
  * password: password for BasicAuth (optional)
  * adapter: Tesla adapter config
  * middleware: List of additional middleware module config (optional)

  ## Examples

      iex> client = ConfluentSchemaRegistry.client()
      %Tesla.Client{
        adapter: nil,
        fun: nil,
        post: [],
        pre: [
          {Tesla.Middleware.BaseUrl, :call, ["http://localhost:8081"]},
          {Tesla.Middleware.Headers, :call,
           [[{"content-type", "application/vnd.schemaregistry.v1+json"}]]},
          {Tesla.Middleware.JSON, :call,
           [[decode_content_types: ["application/vnd.schemaregistry.v1+json"]]]}
        ]
      }
  """
  @spec client(Keyword.t()) :: Tesla.Client.t()
  def client(opts \\ []) do
    base_url = opts[:base_url] || "http://localhost:8081"
    opts_middleware = opts[:middleware] || []
    adapter = opts[:adapter]

    middleware =
      opts_middleware ++
        [
          {Tesla.Middleware.BaseUrl, base_url},
          {Tesla.Middleware.Headers,
           [
             {"content-type", "application/vnd.schemaregistry.v1+json"},
             {"accept",
              "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json"}
           ]},
          {Tesla.Middleware.JSON,
           decode_content_types: [
             "application/vnd.schemaregistry.v1+json",
             "application/vnd.schemaregistry+json"
           ]}
        ] ++ basic_auth(opts)

    Tesla.client(middleware, adapter)
  end

  # Configure Tesla.Middleware.BasicAuth
  @spec basic_auth(Keyword.t()) :: [{Tesla.Middleware.BasicAuth, map}]
  defp basic_auth(opts) do
    if opts[:username] do
      auth_opts =
        opts
        |> Keyword.take([:username, :password])
        |> Map.new()

      [{Tesla.Middleware.BasicAuth, auth_opts}]
    else
      []
    end
  end

  @doc ~S"""
  Get the schema string identified by the input ID.

  https://docs.confluent.io/current/schema-registry/develop/api.html#get--schemas-ids-int-%20id

  Returns binary schema.

  ## Examples

      iex> ConfluentSchemaRegistry.get_schema(client, 21)
      {:ok, "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}"}

  """
  @spec get_schema(Tesla.Client.t(), id) :: {:ok, schema} | {:error, code, reason}
  def get_schema(client, id) when is_integer(id) do
    case do_get(client, "/schemas/ids/#{id}") do
      {:ok, %{"schema" => value}} -> {:ok, value}
      {:ok, value} -> {:error, 1, "Unexpected response: " <> value}
      error -> error
    end
  end

  @doc ~S"""
  Get a list of registered subjects.

  https://docs.confluent.io/current/schema-registry/develop/api.html#get--subjects

  Returns list of subject name strings.

  ## Examples

      iex> ConfluentSchemaRegistry.get_subjects(client)
      {:ok, ["test"]}

  """
  @spec get_subjects(Tesla.Client.t()) :: {:ok, list(subject)} | {:error, code, reason}
  def get_subjects(client) do
    case do_get(client, "/subjects") do
      {:ok, _} = result -> result
      error -> error
    end
  end

  @doc ~S"""
  Get a list of versions registered under the specified subject.

  https://docs.confluent.io/current/schema-registry/develop/api.html#get--subjects-(string-%20subject)-versions

  ```
  {:ok, [1, 2, 3, 4]} = ConfluentSchemaRegistry.get_versions(client, "test")
  ```

  Returns list of integer ids.
  """
  @spec get_versions(Tesla.Client.t(), subject) :: {:ok, list(id)} | {:error, code, reason}
  def get_versions(client, subject) do
    case do_get(client, "/subjects/#{subject}/versions") do
      {:ok, _} = result -> result
      error -> error
    end
  end

  @doc ~S"""
  Deletes the specified subject and its associated compatibility level if registered.
  It is recommended to use this API only when a topic needs to be recycled or
  in development environment.

  https://docs.confluent.io/current/schema-registry/develop/api.html#delete--subjects-(string-%20subject)

  Returns list of integer ids.

  ## Examples

      iex> ConfluentSchemaRegistry.delete_subject(client, "test")
      {:ok, [1]}

  """
  @spec delete_subject(Tesla.Client.t(), subject) :: {:ok, list(id)} | {:error, code, reason}
  def delete_subject(client, subject) do
    case do_delete(client, "/subjects/#{subject}") do
      {:ok, _} = result -> result
      error -> error
    end
  end

  @doc ~S"""
  Get a specific version of the schema registered under this subject

  https://docs.confluent.io/current/schema-registry/develop/api.html#delete--subjects-(string-%20subject)

  Returns a map with the following keys:

  * subject (string) -- Name of the subject that this schema is registered under
  * id (int) -- Globally unique identifier of the schema
  * version (int) -- Version of the returned schema
  * schema (string) -- The Avro schema string

  ```
  case ConfluentSchemaRegistry.get_schema(client, "test", "latest") do
    {:ok, reg} ->
      # Already registered
      schema = reg["schema"]
      schema_id = reg["id"]
    {:error, 404, %{"error_code" => 40401}} ->
      # Subject not found
    {:error, 404, %{"error_code" => 40402}} ->
      # Version not found
    {:error, 422, reason} ->
      # Unprocessable Entity, Invalid Avro version
    {:error, code, reason} ->
      # Other error
  end
  ```

  ## Examples

      iex> ConfluentSchemaRegistry.get_schema(client, "test")
      {:ok,
       %{
         "id" => 21,
         "schema" => "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}",
         "subject" => "test",
         "version" => 13
       }}

      iex> ConfluentSchemaRegistry.get_schema(client, "test", 13)
      {:ok,
       %{
         "id" => 21,
         "schema" => "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}",
         "subject" => "test",
         "version" => 13
       }}

  """
  @spec get_schema(Tesla.Client.t(), subject, version) ::
          {:ok, map} | {:error, code, reason}
  def get_schema(client, subject, version \\ "latest") do
    case do_get(client, "/subjects/#{subject}/versions/#{version}") do
      {:ok, _} = result -> result
      error -> error
    end
  end

  # NOTE: /subjects/#{subject}/versions/#{version}/schema not implemented, as
  # it's redundant with get_schema/3

  @doc ~S"""
  Register a new schema under the specified subject. If successfully
  registered, this returns the unique identifier of this schema in the registry.

  https://docs.confluent.io/current/schema-registry/develop/api.html#post--subjects-(string-%20subject)-versions

  Returns the integer id.

  ```
  case ConfluentSchemaRegistry.register_schema(client, "test", schema) do
    {:ok, schema_id} ->
      # Already registered
    {:error, 409, reason} ->
      # Conflict -- Incompatible Avro schema
    {:error, 422, reason} ->
      # Unprocessable Entity, Invalid Avro schema
    {:error, code, reason} ->
      # Other error
  end
  ```

  ## Examples

      iex> ConfluentSchemaRegistry.register_schema(client, "test", schema)
      {:ok, 21}

  """
  @spec register_schema(Tesla.Client.t(), subject, schema) :: {:ok, id} | {:error, code, reason}
  def register_schema(client, subject, schema) do
    case do_post(client, "/subjects/#{subject}/versions", %{schema: schema}) do
      {:ok, %{"id" => value}} -> {:ok, value}
      {:ok, value} -> {:error, 1, "Unexpected response: #{inspect(value)}"}
      error -> error
    end
  end

  @doc ~S"""
  Check if a schema has already been registered under the specified subject. If
  so, this returns the schema string along with its globally unique identifier,
  its version under this subject and the subject name.

  https://docs.confluent.io/current/schema-registry/develop/api.html#post--subjects-(string-%20subject)

  Returns map with the following keys:

  * subject (string) -- Name of the subject that this schema is registered under
  * id (int) -- Globally unique identifier of the schema
  * version (int) -- Version of the returned schema
  * schema (string) -- The Avro schema string

  ```
  schema = "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}"
  case ConfluentSchemaRegistry.is_registered(client, "test", schema) do
    {:ok, reg} ->
      # Found
      schema = reg["schema"]
      schema_id = reg["id"]
    {:error, 404, %{"error_code" => 40401}} ->
      # Subject not found
    {:error, 404, %{"error_code" => 40403}} ->
      # Schema not found
    {:error, code, reason} ->
      # Other error
  end
  ```

  ## Examples

      iex> ConfluentSchemaRegistry.is_registered(client, "test", schema)
      {:ok,
       %{
         "id" => 21,
         "schema" => "{\"type\":\"record\",\"name\":\"test\",\"fields\":[{\"name\":\"field1\",\"type\":\"string\"},{\"name\":\"field2\",\"type\":\"int\"}]}",
         "subject" => "test",
         "version" => 1
       }}

      iex> ConfluentSchemaRegistry.is_registered(client, "test2", schema)
      {:error, 404, %{"error_code" => 40401, "message" => "Subject not found. ..."}}

  """
  @spec is_registered(Tesla.Client.t(), subject, schema) :: {:ok, map} | {:error, code, reason}
  def is_registered(client, subject, schema) do
    case do_post(client, "/subjects/#{subject}", %{schema: schema}) do
      {:ok, _} = result -> result
      error -> error
    end
  end

  @doc ~S"""
  Deletes a specific version of the schema registered under this subject. This
  only deletes the version and the schema ID remains intact making it still
  possible to decode data using the schema ID.

  https://docs.confluent.io/current/schema-registry/develop/api.html#delete--subjects-(string-%20subject)-versions-(versionId-%20version)

  ```
  {:ok, 1} = ConfluentSchemaRegistry.delete_version(client, "test") # latest
  {:ok, 1} = ConfluentSchemaRegistry.delete_version(client, "test", 1)
  ```

  Returns integer id of deleted version.
  """
  @spec delete_version(Tesla.Client.t(), subject, version) :: {:ok, id} | {:error, code, reason}
  def delete_version(client, subject, version \\ "latest") do
    case do_delete(client, "/subjects/#{subject}/versions/#{version}") do
      {:ok, value} when is_integer(value) -> {:ok, value}
      {:ok, value} -> {:error, 1, "Unexpected response: #{inspect(value)}"}
      error -> error
    end
  end

  @doc ~S"""
  Test input schema against a particular version of a subject's schema for
  compatibility. Note that the compatibility level applied for the check is the
  configured compatibility level for the subject (`get_compatibility/2`).
  If this subject's compatibility level was never changed, then the
  global compatibility level applies (`get_compatibility/1`).

  https://docs.confluent.io/current/schema-registry/develop/api.html#post--compatibility-subjects-(string-%20subject)-versions-(versionId-%20version)

  Returns boolean.

  ## Examples

      iex> ConfluentSchemaRegistry.is_compatible(client, "test", schema)
      {:ok, true}

      iex> ConfluentSchemaRegistry.is_compatible(client, "test", schema, "latest")
      {:ok, true}

      iex> ConfluentSchemaRegistry.is_compatible(client, "test", schema, 1)
      {:ok, true}

  """
  @spec is_compatible(Tesla.Client.t(), subject, schema, version) ::
          {:ok, boolean} | {:error, code, reason}
  def is_compatible(client, subject, schema, version \\ "latest") do
    case do_post(client, "/compatibility/subjects/#{subject}/versions/#{version}", %{
           schema: schema
         }) do
      {:ok, %{"is_compatible" => value}} -> {:ok, value}
      {:ok, value} -> {:error, 1, "Unexpected response: " <> value}
      error -> error
    end
  end

  @doc ~S"""
  Update global compatibility level.

  Level is a string which must be one of BACKWARD, BACKWARD_TRANSITIVE, FORWARD,
  FORWARD_TRANSITIVE, FULL, FULL_TRANSITIVE, NONE

  https://docs.confluent.io/current/schema-registry/develop/api.html#put--config

  Returns string.

  ## Examples

      iex> ConfluentSchemaRegistry.update_compatibility(client, "test", "FULL")
      {:ok, "FULL"}

  """
  @spec update_compatibility(Tesla.Client.t(), level) :: {:ok, level} | {:error, code, reason}
        when level: binary, code: non_neg_integer, reason: any
  def update_compatibility(client, level) do
    case do_put(client, "/config", %{compatibility: level}) do
      {:ok, %{"compatibility" => value}} -> {:ok, value}
      {:ok, value} -> {:error, 1, "Unexpected response: " <> value}
      error -> error
    end
  end

  @doc ~S"""
  Get global compatibility level.

  Level is a string which will be one of BACKWARD, BACKWARD_TRANSITIVE, FORWARD,
  FORWARD_TRANSITIVE, FULL, FULL_TRANSITIVE, NONE

  https://docs.confluent.io/current/schema-registry/develop/api.html#put--config

  Returns string.

  ## Examples

      iex> ConfluentSchemaRegistry.get_compatibility(client)
      {:ok, "BACKWARD"}

  """
  @spec get_compatibility(Tesla.Client.t()) :: {:ok, level} | {:error, code, reason}
        when level: binary, code: non_neg_integer, reason: any
  def get_compatibility(client) do
    case do_get(client, "/config") do
      {:ok, %{"compatibilityLevel" => value}} -> {:ok, value}
      {:ok, value} -> {:error, 1, "Unexpected response: #{inspect(value)}"}
      error -> error
    end
  end

  @doc ~S"""
  Update compatibility level for the specified subject.

  Leve is a string which must be one of BACKWARD, BACKWARD_TRANSITIVE, FORWARD,
  FORWARD_TRANSITIVE, FULL, FULL_TRANSITIVE, NONE

  https://docs.confluent.io/current/schema-registry/develop/api.html#put--config

  Returns string.

  ## Examples

      iex> ConfluentSchemaRegistry.update_compatibility(client, "test", "FULL")
      {:ok, "FULL"}

  """
  @spec update_compatibility(Tesla.Client.t(), subject, level) ::
          {:ok, level} | {:error, code, reason}
  def update_compatibility(client, subject, level) do
    case do_put(client, "/config/#{subject}", %{compatibility: level}) do
      {:ok, %{"compatibility" => value}} -> {:ok, value}
      {:ok, value} -> {:error, 1, "Unexpected response: #{inspect(value)}"}
      error -> error
    end
  end

  @doc ~S"""
  Get compatibility level for a subject.

  Level is a string which will be one of BACKWARD, BACKWARD_TRANSITIVE, FORWARD,
  FORWARD_TRANSITIVE, FULL, FULL_TRANSITIVE, NONE

  https://docs.confluent.io/current/schema-registry/develop/api.html#put--config

  Returns string.

  ## Examples

      iex> ConfluentSchemaRegistry.get_compatibility(client, "test")
      {:ok, "FULL"}

  """
  @spec get_compatibility(Tesla.Client.t(), subject) ::
          {:ok, level} | {:error, code, reason}
  def get_compatibility(client, subject) do
    case do_get(client, "/config/#{subject}") do
      {:ok, %{"compatibilityLevel" => value}} -> {:ok, value}
      {:ok, value} -> {:error, 1, "Unexpected response: " <> value}
      error -> error
    end
  end

  # Internal utility functions

  @spec do_get(Tesla.Client.t(), binary) :: {:ok, any} | {:error, code, reason}
  defp do_get(client, url) do
    tesla_response(Tesla.get(client, url))
  end

  @spec do_delete(Tesla.Client.t(), binary) :: {:ok, any} | {:error, code, reason}
  defp do_delete(client, url) do
    tesla_response(Tesla.delete(client, url))
  end

  @spec do_post(Tesla.Client.t(), binary, any) :: {:ok, any} | {:error, code, reason}
  defp do_post(client, url, data) when is_binary(data) do
    tesla_response(Tesla.post(client, url, data))
  end

  defp do_post(client, url, data) do
    case Jason.encode(data) do
      {:ok, encoded} ->
        do_post(client, url, encoded)

      {:error, reason} ->
        {:error, 0, reason}
    end
  end

  @spec do_put(Tesla.Client.t(), binary, any) :: {:ok, any} | {:error, code, reason}
  defp do_put(client, url, data) when is_binary(data) do
    tesla_response(Tesla.put(client, url, data))
  end

  defp do_put(client, url, data) do
    case Jason.encode(data) do
      {:ok, encoded} ->
        do_put(client, url, encoded)

      {:error, reason} ->
        {:error, 0, reason}
    end
  end

  defp tesla_response({:ok, %{status: 200, body: body}}), do: {:ok, body}
  defp tesla_response({:ok, %{status: status, body: body}}), do: {:error, status, body}
  defp tesla_response({:error, reason}), do: {:error, 0, reason}
end
